File: solutions/cpp/secret-handshake/secret_handshake.cpp
#include "secret_handshake.h"

#include <algorithm>

namespace secret_handshake
{
    using namespace std;

    vector<string> commands(const int code)
    {
        vector<string> handshake;
        auto secret = [&handshake, code](int mask, string action)
        {
            if ((code & mask) == 0)
                return;
            if (action == "reverse")
                reverse(handshake.begin(), handshake.end());
            else
                handshake.emplace_back(action);
        };

        secret(0b00001, "wink");
        secret(0b00010, "double blink");
        secret(0b00100, "close your eyes");
        secret(0b01000, "jump");
        secret(0b10000, "reverse");
        return handshake;
    }

} // namespace secret_handshake

File: solutions/elixir/secret-handshake/lib/secret_handshake.ex
defmodule SecretHandshake do
  import Bitwise

  @doc """
  Determine the actions of a secret handshake based on the binary
  representation of the given `code`.

  If the following bits are set, include the corresponding action in your list
  of commands, in order from lowest to highest.

  1 = wink
  10 = double blink
  100 = close your eyes
  1000 = jump

  10000 = Reverse the order of the operations in the secret handshake
  """
  @spec commands(code :: integer) :: list(String.t())
  def commands(code) do
    []
    |> secret(band(code, 0b01000) == 0b01000, "jump")
    |> secret(band(code, 0b00100) == 0b00100, "close your eyes")
    |> secret(band(code, 0b00010) == 0b00010, "double blink")
    |> secret(band(code, 0b00001) == 0b00001, "wink")
    |> secret(band(code, 0b10000) == 0b10000, &Enum.reverse/1)
  end

  defp secret(handshake, do?, todo)
  defp secret(handshake, false, _), do: handshake
  defp secret(handshake, true, str) when is_binary(str), do: [str, handshake]
  defp secret(handshake, true, todo), do: fun.(handshake)
end
